### DISCUSSION TITLE
How Do I Securely Connect Two Tables with Auth and RLS in Supabase? Best Practices for Table Architecture and Relationships?

### DISCUSSION THREAD

**Discussion by vladimirzb** [supabase/supabase#17115] [Questions] [answered]
### Question:

I'm working on a project using Supabase and have two tables, `Ingresos` and `FuenteIngreso`. Currently, these tables are not connected, but I'm considering creating a relationship between them. Both tables have a relationship with the `auth` table to facilitate authentication.

Here's a brief overview of the tables:

- `Ingresos`: Stores information about various income sources.
- `FuenteIngreso`: Stores the types or categories of income.

My questions are:

1. What is the most secure way to establish a relationship between `Ingresos` and `FuenteIngreso`?
2. Should I remove the relationship between `FuenteIngreso` and `auth` if I connect it with `Ingresos`, which already has a relationship with `auth`?
3. Does removing the relationship between `FuenteIngreso` and `auth` pose any security risks, given that `Ingresos` is already connected to `auth`?

I'm particularly interested in understanding how to best use Row-Level Security (RLS) in this context.

Any insights or best practices would be greatly appreciated.

**Comment by GaryAustin1** [comment, marked as answer]
I answered this in Discord also.
Having a foreign key to auth really has little to do with RLS security.  That is just a way to possibly cascade delete and enforce a user id exists.  You could have a user_id column not linked at all and as long as it is a valid UUID you can use RLS on it.  When you insert or update you would have a policy with auth.uid(). This is the security that ensures the user_id column is valid.

If you are planning to use foreign keys to join tables in the API, you need to really have a public schema users or profile "top" table that is your single fk link to auth.users.  Then your other tables all fk to that table.  The API cannot do  joins thru the auth schema.

**Reply by vladimirzb** [reply]
Thanks for the help! I am going to summarize what we discussed on [Discord](https://discord.com/channels/839993398554656828/1147536725711802409/1147536725711802409) and GitHub.

I wanted to share the database structure and relations I've settled on for handling authentication in Supabase. This might be useful for anyone looking to set up a similar architecture.

#### Database Tables:

1. **Profiles**: This table serves as the primary table and is linked to `auth.users`. It has a unique constraint on `user_id`, which is a foreign key referencing `auth.users`.

2. **FuenteIngreso**: This table stores types or categories of income and has a foreign key `user_id` that references the `Profiles` table.

3. **Gastos**: This table stores expenses and also has a foreign key `user_id` that references the `Profiles` table.

4. **Ingresos**: This table stores income and has two foreign keys: one `user_id` that references the `Profiles` table and another `fuente` that references the `FuenteIngreso` table.

#### SQL Schema Code:

```sql
-- Profiles Table
create table public."Profiles" (
  id bigint generated by default as identity,
  user_id uuid not null default auth.uid (),
  constraint Profiles_pkey primary key (id),
  constraint Profiles_user_id_key unique (user_id),
  constraint Profiles_user_id_fkey foreign key (user_id) references auth.users (id) on delete cascade
);

-- FuenteIngreso Table
create table public."FuenteIngreso" (
  id bigint generated by default as identity,
  created_at timestamp with time zone not null default now(),
  fuente_name text not null,
  user_id uuid not null,
  constraint FuenteIngreso_pkey primary key (id),
  constraint FuenteIngreso_user_id_key unique (user_id),
  constraint FuenteIngreso_user_id_fkey foreign key (user_id) references "Profiles" (user_id) on delete cascade
);

-- gastos Table
create table public.gastos (
  id bigint generated always as identity,
  fuente text not null,
  cantidad numeric not null,
  timestamp_with_time_zone timestamp with time zone not null,
  user_id uuid not null default auth.uid (),
  constraint gastos_pkey primary key (id),
  constraint gastos_user_id_key unique (user_id),
  constraint gastos_user_id_fkey foreign key (user_id) references "Profiles" (user_id) on delete cascade
);

-- ingresos Table
create table public.ingresos (
  id bigint generated always as identity,
  fuente bigint not null,
  cantidad numeric not null,
  timestamp_with_time_zone timestamp with time zone not null,
  user_id uuid not null default auth.uid (),
  constraint ingresos_pkey primary key (id),
  constraint ingresos_user_id_key unique (user_id),
  constraint ingresos_fuente_fkey foreign key (fuente) references "FuenteIngreso" (id) on delete cascade,
  constraint ingresos_user_id_fkey foreign key (user_id) references "Profiles" (user_id) on delete cascade
);

```

#### Row-Level Security Policies:

For each table, I've set up RLS policies to restrict access based on the `user_id`. Here's an example for the `Profiles` table:

```sql
CREATE POLICY profiles_user_policy
  FOR SELECT ON profiles
  USING (user_id = auth.uid());
```

You can create similar policies for other tables, replacing the table name and conditions as needed.

I hope this helps anyone looking to set up a similar structure!

Best regards,
Vladimir
